synapseclient/api/entity_factory.py

```diff
@@ -0,0 +1,391 @@
+            calculated.
+        limit_search: Limit the search to a specific project or folder.
+        synapse_client: If not passed in or None this will use the last client from
+            the `.login()` method.
+
+    Returns:
+        A dictionary containing the entity bundle of the file found.
+
+    Raises:
+        SynapseFileNotFoundError: If the file is not found in Synapse.
+    """
+    from synapseclient import Synapse
+
+    syn = Synapse.get_client(synapse_client=synapse_client)
+    md5 = md5 or utils.md5_for_file_hex(filename=filepath)
+    results = (
+        await get_entities_by_md5(
+            md5=md5,
+            synapse_client=synapse_client,
+        )
+    )["results"]
+    if limit_search is not None:
+        # Go through and find the path of every entity found
+        paths = [
+            await get_entity_path(entity_id=ent["id"], synapse_client=synapse_client)
+            for ent in results
+        ]
+        # Filter out all entities whose path does not contain limitSearch
+        results = [
+            ent
+            for ent, path in zip(results, paths)
+            if utils.is_in_path(id=limit_search, path=path)
+        ]
+    if len(results) == 0:  # None found
+        raise SynapseFileNotFoundError(f"File {filepath} not found in Synapse")
+    elif len(results) > 1:
+        id_txts = "\n".join([f"{r['id']}.{r['versionNumber']}" for r in results])
+        syn.logger.warning(
+            f"\nThe file {filepath} is associated with many files in Synapse:\n{id_txts}\n"
+            "You can limit to files in specific project or folder by setting the limitSearch to the"
+            " synapse Id of the project or folder.\n"
+            "Will use the first one returned: \n"
+            f"{results[0]['id']} version {results[0]['versionNumber']}\n"
+        )
+    entity_id = results[0]["id"]
+    entity_version = results[0]["versionNumber"]
+
+    bundle = await get_entity_id_version_bundle2(
+        entity_id=entity_id,
+        version=entity_version,
+        synapse_client=synapse_client,
+    )
+    syn.cache.add(
+        file_handle_id=bundle["entity"]["dataFileHandleId"], path=filepath, md5=md5
+    )
+
+    return bundle
+
+
+async def _cast_into_class_type(
+    entity_bundle: Dict[str, Any],
+    download_file: bool = True,
+    download_location: str = None,
+    if_collision: str = None,
+    submission: str = None,
+    follow_link: bool = False,
+    entity_to_update: Union["Project", "File", "Folder"] = None,
+    *,
+    synapse_client: Optional["Synapse"] = None,
+) -> Union["Project", "File", "Folder"]:
+    """
+    Take an entity_bundle returned from the Synapse API and cast it into the appropriate
+    class type. This will also download the file if `download_file` is set to True.
+    Additionally, if `entity_to_update` is passed in, the fields of the existing entity
+    will be updated instead of creating a new instance. If the entity is a link and
```

Comment:
nit: "If the entity is a link..." but links are not supported yet. I worry this will lead to confusion, unless you are planning on including Links pretty soon

Reply:
Links are supported enough to be included here. We are not returning a Link entity yet, however, we are supporting following a link to find a target entity.

For example links that look like the following are supported:

1. Link -> Project
2. Link -> Folder
3. Link -> File

If link is not pointing to one of those resources a `ValueError` is raised.

Reply:
Makes sense! Thanks